posthog/heatmaps/sql.py

```diff
@@ -0,0 +1,163 @@
+from posthog.clickhouse.table_engines import (
+    Distributed,
+    ReplicationScheme,
+    MergeTreeEngine,
+)
+from posthog.kafka_client.topics import KAFKA_CLICKHOUSE_HEATMAP_EVENTS
+
+HEATMAPS_DATA_TABLE = lambda: "sharded_heatmaps"
+
+
+"""
+We intend to send specific $heatmap events to build a heatmap instead of building from autocapture like the click map
+We'll be storing indiviudal clicks per url/team/session
+And we'll be querying for those clicks at day level of granularity
+And we'll be querying by URL exact or wildcard match
+And we'll _sometimes_ be querying by width
+
+We _could_ aggregate this data by day, but we're hoping this will be small/fast enough not to bother
+And can always add a materialized view for day (and week?) granularity driven by this data if needed
+
+We only add session_id so that we could offer example sessions for particular clicked areas in the toolbar
+"""
+
+KAFKA_HEATMAPS_TABLE_BASE_SQL = """
+CREATE TABLE IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
+(
+    $session_id VARCHAR,
+    team_id Int64,
+    timestamp DateTime64(6, 'UTC'),
+    -- x is the x with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
+    x Int16,
+    -- y is the y with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
+    y Int16,
+    -- stored so that in future we can support other resolutions
+    scale_factor Int16,
+    $viewport_width Int16,
+    $viewport_height Int16,
+    -- some elements move when the page scrolls, others do not
+    $pointer_target_fixed Bool,
+    $current_url VARCHAR
+) ENGINE = {engine}
+"""
+
+HEATMAPS_TABLE_BASE_SQL = """
+CREATE TABLE IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
+(
+    $session_id VARCHAR,
+    team_id Int64,
+    timestamp DateTime64(6, 'UTC'),
+    -- x is the x with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
+    x Int16,
+    -- y is the y with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
+    y Int16,
+    -- stored so that in future we can support other resolutions
+    scale_factor Int16,
+    $viewport_width Int16,
+    $viewport_height Int16,
+    -- some elements move when the page scrolls, others do not
+    $pointer_target_fixed Bool,
+    $current_url VARCHAR,
+    _timestamp DateTime
+) ENGINE = {engine}
+"""
+
+HEATMAPS_DATA_TABLE_ENGINE = lambda: MergeTreeEngine("heatmaps", replication_scheme=ReplicationScheme.SHARDED)
+
+HEATMAPS_TABLE_SQL = lambda: (
+    HEATMAPS_TABLE_BASE_SQL
+    + """
+    PARTITION BY toYYYYMM(timestamp)
+    -- almost always this is being queried by
+    --   * team_id,
+    --   * date range,
+    --   * URL (maybe matching wild cards),
+    --   * width
+    -- we'll almost never query this by session id
+    -- so from least to most cardinality that's
+    ORDER BY (team_id, $viewport_width, $current_url, toDate(timestamp))
```

Comment:
Honestly in this case, since you are querying it by `team_id` and `date` most of the time you might want to swap this around a bit. The cardinality is important, but the frequency of hitting these in the predicate is also important. `toDate(` helps a lot with the cardinality too. I don't really know which is better for the last two between `$current_url` and `$viewport_width` so feel free to ignore me there.
```suggestion
    ORDER BY (team_id,  toDate(timestamp), $current_url, $viewport_width)
```